src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp

```diff
@@ -0,0 +1,610 @@
+    QObject::tr("B2 Reverse"),
+    QObject::tr("B3 Forward"),
+    QObject::tr("B3 Reverse")
+};
+
+PCRPrimerDesignForDNAAssemblyTask::PCRPrimerDesignForDNAAssemblyTask(const PCRPrimerDesignForDNAAssemblyTaskSettings& _settings, const QByteArray& _sequence)
+    : Task("PCR Primer Design For DNA Assembly Task", TaskFlags_FOSCOE | TaskFlag_ReportingIsSupported | TaskFlag_ReportingIsEnabled),
+      settings(_settings),
+      sequence(_sequence),
+      reverseComplementSequence(DNASequenceUtils::reverseComplement(sequence)) {
+    GCOUNTER(cvar, "PCRPrimerDesignForDNAAssemblyTask");
+
+    tpm = ProgressManagement::Progress_Manual;
+}
+
+void PCRPrimerDesignForDNAAssemblyTask::prepare() {
+    auto prepareLoadDocumentTask = [this](const QString& url) -> LoadDocumentTask* {
+        QList<FormatDetectionResult> formats = DocumentUtils::detectFormat(url);
+        CHECK_EXT(!formats.isEmpty(), setError(tr("Unknown file format!")), nullptr);
+
+        auto ioRegistry = AppContext::getIOAdapterRegistry();
+        SAFE_POINT_EXT(ioRegistry != nullptr, setError(L10N::nullPointerError("IOAdapterRegistry")), nullptr);
+
+        IOAdapterFactory* iow = ioRegistry->getIOAdapterFactoryById(BaseIOAdapters::LOCAL_FILE);
+        return new LoadDocumentTask(formats.first().format->getFormatId(), GUrl(url), iow);
+    };
+
+    if (!settings.backboneSequenceUrl.isEmpty()) {
+        loadBackboneSequence = prepareLoadDocumentTask(settings.backboneSequenceUrl);
+        CHECK_OP(stateInfo, );
+
+        addSubTask(loadBackboneSequence);
+    }
+
+    if (!settings.otherSequencesInPcrUrl.isEmpty()) {
+        loadOtherSequencesInPcr = prepareLoadDocumentTask(settings.otherSequencesInPcrUrl);
+        CHECK_OP(stateInfo, );
+
+        addSubTask(loadOtherSequencesInPcr);
+    }
+
+    findUnwantedIslands = new FindUnwantedIslandsTask(settings.leftArea, settings.overlapLength.maxValue, sequence, false);
+    addSubTask(findUnwantedIslands);
+
+    U2Region reverseComplementArea = DNASequenceUtils::reverseComplementRegion(settings.rightArea, reverseComplementSequence.size());
+    findUnwantedIslandsReverseComplement = new FindUnwantedIslandsTask(reverseComplementArea, settings.overlapLength.maxValue, reverseComplementSequence, true);
+    addSubTask(findUnwantedIslandsReverseComplement);
+
+
+}
+
+void PCRPrimerDesignForDNAAssemblyTask::run() {
+    // We have A and B1, B2, B3 pairs of primers
+    // A and B1 are pretty the same, except the fact that B1 has some depencedcies on B2 and B3
+    // So, the first time we find good B1 pair they are also the A pair. But if B1 doesn't fit to B2 or B3, we need to continue searching B1, which will be fit to B2 and B3
+    bool aWasNotFoundYet = true;
+    for (const auto& regionBetweenIslandsForward : regionsBetweenIslandsForward) {
+        stateInfo.setProgress(((double)regionsBetweenIslandsForward.indexOf(regionBetweenIslandsForward) / (double)regionsBetweenIslandsForward.size()) * 100);
```

Comment:
Really? Why not using indexed for loop instead?

Reply:
Got it